_drafts/cortex-m-breakpoints.md

```diff
@@ -0,0 +1,607 @@
+
+With that out of the way, let's take a look at the actual **GDB RSP** commands that get sent over the wire when we install a breakpoint. Conveniently GDB has a builtin debug command, `set debug remote 1`, which can be used to enable printing of every single command sent between the client and server!
+
+To facilitate breakpoint testing, I've created some test functions in the demo app `dummy_function_*` that we can install breakpoints on. We can use the `dump_dummy_funcs` CLI command to print information about them all:
+
+```
+shell> dump_dummy_funcs
+dummy_function_1: Starts at 0x240. First Instruction = 0x48024901
+dummy_function_2: Starts at 0x250. First Instruction = 0x48024901
+dummy_function_3: Starts at 0x260. First Instruction = 0x48024901
+dummy_function_4: Starts at 0x270. First Instruction = 0x48024901
+dummy_function_5: Starts at 0x280. First Instruction = 0x48024901
+dummy_function_6: Starts at 0x290. First Instruction = 0x48024901
+dummy_function_7: Starts at 0x2a0. First Instruction = 0x48024901
+dummy_function_8: Starts at 0x2b0. First Instruction = 0x48024901
+dummy_function_9: Starts at 0x2c0. First Instruction = 0x48024901
+dummy_function_ram: Starts at 0x20000000. First Instruction = 0x48024901
+```
+
+#### GDB RSP while Setting Breakpoints
+
+```bash
+(gdb) set debug remote 1
+(gdb) break dummy_function_1
+Sending packet: $m80,4#35...Packet received: 2de9f341
+Sending packet: $m240,40#93...Packet received:[...]
+Sending packet: $m248,4#6b...Packet received: 2f730000
+Sending packet: $m248,4#6b...Packet received: 2f730000
+Sending packet: $m240,2#61...Packet received: 0149
+Breakpoint 1 at 0x240: file stub_functions.c, line 4.
+```
+
+We see some [`m` packets](https://sourceware.org/gdb/onlinedocs/gdb/Packets.html#index-m-packet) (`m addr,length`). These are used for requesting reads of memory but no breakpoint commands?! Let's `continue` and see what happens:
+
+```bash
+(gdb) continue
+Continuing.
+(gdb) c
+Continuing.
+==> Sending packet: $Z0,240,2#aa...Packet received: OK
+==> Packet Z0 (software-breakpoint) is supported
+Sending packet: $vCont?#49...Packet received:
+Packet vCont (verbose-resume) is NOT supported
+Sending packet: $Hc0#db...Packet received: OK
+Sending packet: $c#63...
+```
+
+We see a `$Z0,240,2` request issued which means "install a software breakpoint at address 0x240". This matches the address of `dummy_function_1` so that's what we'd expect! It looks like breakpoints are only installed when we resume.
+
+#### GDB RSP when a Breakpoint Triggers
+
+Now let's actually cause the breakpoint to be triggered by executing the function we installed the breakpoint on. This can be done by running the `call_dummy_funcs` CLI command:
+
+```
+shell> call_dummy_funcs
+[... GDB should have halted ...]
+```
+
+```bash
+Sending packet: $c#63...Packet received: T05hwbreak:;thread:0000DEAD;
+Sending packet: $g#67...Packet received:[...]
+Sending packet: $qfThreadInfo#bb...Packet received: m0000dead
```

Comment:
A note on this `0000DEAD` address would be interesting

Reply:
I added a little note about it. It is interesting hehe
<img width="727" alt="image" src="https://user-images.githubusercontent.com/7494333/84837839-08602600-b007-11ea-8454-4d8652f99b51.png">